pkg/dockerregistry/server/localblobstore.go

```diff
@@ -0,0 +1,139 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	"k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type localBlobStore struct {
+	distribution.BlobStore
+
+	repo *repository
+}
+
+var _ distribution.BlobStore = &localBlobStore{}
+
+func (bs *localBlobStore) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
+	desc, err := bs.BlobStore.Stat(ctx, dgst)
+	switch {
+	case err == distribution.ErrBlobUnknown:
+		// continue on to the code below and look up the blob in a local store
+	case err != nil:
+		context.GetLogger(bs.repo.ctx).Errorf("Failed to find blob %s: %#v", dgst.String(), err)
+		fallthrough
+	default:
+		return desc, err
+	}
+
+	is, err := bs.repo.getImageStream()
+	if err != nil {
+		if errors.IsNotFound(err) || errors.IsForbidden(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		context.GetLogger(bs.repo.ctx).Errorf("Error retrieving image stream for blob: %s", err)
+		return distribution.Descriptor{}, err
+	}
+
+	cached := bs.repo.cachedLayers.RepositoriesForDigest(dgst)
+
+	// look at the first level of tagged repositories first
+	search := bs.identifyCandidateRepositories(is, true)
```

Comment:
No. There is no way to get the App out of context. 

Context in dispatcher looks like this: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/handlers/app.go#L693

here we make it : https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/handlers/app.go#L594

Here middleware creation: 
* registerMiddleware: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/handlers/app.go#L226
* repoMiddleware: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/handlers/app.go#L628

In both cases we get the unusable (for us) context.